File: src/core/schedulingEngine.ts
import {
  AutoScheduleResult,
  AutoSchedulePreviewItem,
  SkippedDateItem,
  DayCapacitySummary,
  DailyCapacitySetting,
  Session,
  SchedulePattern
} from '../types';

export const DEFAULT_MAX_CAPACITY = 10;

export const ARABIC_DAYS = [
  'الأحد',
  'الإثنين',
  'الثلاثاء',
  'الأربعاء',
  'الخميس',
  'الجمعة',
  'السبت'
];

export const PATTERN_DAYS_MAP: Record<SchedulePattern, number[]> = {
  SAT_MON_WED: [6, 1, 3], // السبت، الإثنين، الأربعاء
  SUN_TUE_THU: [0, 2, 4], // الأحد، الثلاثاء، الخميس
  DAILY: [0, 1, 2, 3, 4, 6], // يومياً عدا الجمعة
  CUSTOM: []
};

/**
 * دالة مساعدة لتنسيق التاريخ بصيغة YYYY-MM-DD
 */
export function formatDate(d: Date): string {
  const year = d.getFullYear();
  const month = String(d.getMonth() + 1).padStart(2, '0');
  const day = String(d.getDate()).padStart(2, '0');
  return `${year}-${month}-${day}`;
}

/**
 * دالة مساعدة للحصول على اسم اليوم بالعربية
 */
export function getArabicDayName(dateStr: string): string {
  const d = new Date(dateStr + 'T00:00:00');
  return ARABIC_DAYS[d.getDay()];
}

/**
 * فحص وتجميع سعة يوم محدد
 */
export function calculateDayCapacity(
  dateStr: string,
  sessions: Session[],
  capacitySettings: Record<string, DailyCapacitySetting>,
  defaultCapacity: number = DEFAULT_MAX_CAPACITY
): DayCapacitySummary {
  const d = new Date(dateStr + 'T00:00:00');
  const dayOfWeek = d.getDay();
  const setting = capacitySettings[dateStr];

  // الجمعة عطلة افتراضية إلا إذا حُدد خلاف ذلك
  const isDefaultWorkingDay = dayOfWeek !== 5; // 5 = الجمعة
  const isWorkingDay = setting !== undefined ? setting.isWorkingDay : isDefaultWorkingDay;
  const maxCapacity = setting !== undefined ? setting.maxCapacity : defaultCapacity;

  // الجلسات المحجوزة التي تستهلك طاقة استيعابية (المجدولة، المكتملة، المؤجلة إلى هذا اليوم)
  // لا نحتسب الجلسات الملغاة (CANCELLED)
  const bookedCount = sessions.filter(
    (s) => s.sessionDate === dateStr && s.status !== 'CANCELLED'
  ).length;

  const percentage = maxCapacity > 0 ? Math.round((bookedCount / maxCapacity) * 100) : 100;

  let status: DayCapacitySummary['status'] = 'available';
  if (!isWorkingDay) {
    status = 'holiday';
  } else if (bookedCount >= maxCapacity) {
    status = 'full';
  } else if (bookedCount >= maxCapacity * 0.8) {
    status = 'near_full';
  }

  return {
    date: dateStr,
    dayNameAr: ARABIC_DAYS[dayOfWeek],
    dayOfWeek,
    bookedCount,
    maxCapacity,
    isWorkingDay,
    percentage,
    status,
    notes: setting?.notes
  };
}

/**
 * محرك التوزيع التلقائي الذكي للجلسات على الأيام الشاغرة
 */
export function autoDistributeSessions(
  startDateStr: string,
  totalSessions: number,
  preferredDays: number[],
  existingSessions: Session[],
  capacitySettings: Record<string, DailyCapacitySetting>,
  allowOverride: boolean = false,
  defaultCapacity: number = DEFAULT_MAX_CAPACITY
): AutoScheduleResult {
  if (totalSessions <= 0) {
    throw new Error('عدد الجلسات يجب أن يكون أكبر من الصفر');
  }

  if (preferredDays.length === 0) {
    throw new Error('يجب تحديد يوم واحد على الأقل في نمط الجدولة');
  }

  const scheduledSessions: AutoSchedulePreviewItem[] = [];
  const skippedDates: SkippedDateItem[] = [];
  
  // نسخة محلية من الجلسات لمحاكاة الحجز المؤقت أثناء التوزيع
  const simulatedSessions = [...existingSessions];

  let currentDate = new Date(startDateStr + 'T00:00:00');
  const MAX_DAYS_LOOKAHEAD = 180; // فحص حتى 6 أشهر كحد أقصى
  let daysScanned = 0;

  while (scheduledSessions.length < totalSessions && daysScanned < MAX_DAYS_LOOKAHEAD) {
    const dayOfWeek = currentDate.getDay();
    const dateStr = formatDate(currentDate);
    const dayNameAr = ARABIC_DAYS[dayOfWeek];

    // هل يطابق النمط المفضل للمريض؟
    if (preferredDays.includes(dayOfWeek)) {
      const cap = calculateDayCapacity(dateStr, simulatedSessions, capacitySettings, defaultCapacity);

      if (!cap.isWorkingDay) {
        skippedDates.push({
          date: dateStr,
          dayNameAr,
          reason: cap.notes ? `عطلة: ${cap.notes}` : 'عطلة أسبوعية أو رسمية'
        });
      } else if (cap.bookedCount >= cap.maxCapacity && !allowOverride) {
        skippedDates.push({
          date: dateStr,
          dayNameAr,
          reason: `اليوم ممتلئ بالكامل (${cap.bookedCount}/${cap.maxCapacity})`
        });
      } else {
        // حجز الجلسة بنجاح
        const isOverrideNeeded = cap.bookedCount >= cap.maxCapacity;
        const sessionNum = scheduledSessions.length + 1;

        scheduledSessions.push({
          sessionNumber: sessionNum,
          date: dateStr,
          dayNameAr,
          dayCapacity: cap.maxCapacity,
          dayBookedCount: cap.bookedCount + 1,
          isOverrideNeeded
        });

        // إضافة جلسة محاكاة لليوم حتى يتأثر الفحص للجلسات اللاحقة في حال كان التكرار في نفس اليوم
        simulatedSessions.push({
          id: `sim-${sessionNum}`,
          planId: 'sim-plan',
          patientId: 'sim-patient',
          sessionNumber: sessionNum,
          sessionDate: dateStr,
          status: 'SCHEDULED',
          createdAt: new Date().toISOString()
        });
      }
    }

    // الانتقال إلى اليوم التالي
    currentDate.setDate(currentDate.getDate() + 1);
    daysScanned++;
  }

  if (scheduledSessions.length < totalSessions) {
    throw new Error(`لم نتمكن من جدولة جميع الجلسات. تم جدولة ${scheduledSessions.length} من أصل ${totalSessions} جلسة.`);
  }

  const endDate = scheduledSessions[scheduledSessions.length - 1].date;

  return {
    scheduledSessions,
    skippedDates,
    endDate
  };
}

/**
 * خوارزمية ترحيل الجلسات المتبقية (Cascade Shift)
 * تقوم بإعادة جدولة الجلسات من جلسة معينة فصاعداً إلى تواريخ شاغرة جديدة
 */
export function shiftSubsequentSessions(
  fromSessionNumber: number,
  newStartDateStr: string,
  preferredDays: number[],
  allPlanSessions: Session[],
  allCenterSessions: Session[],
  capacitySettings: Record<string, DailyCapacitySetting>,
  allowOverride: boolean = false,
  defaultCapacity: number = DEFAULT_MAX_CAPACITY
): { updatedSessions: { sessionId: string; oldDate: string; newDate: string }[]; skippedDates: SkippedDateItem[] } {
  // تصفية الجلسات المستهدفة بالترحيل
  const targetSessions = allPlanSessions
    .filter((s) => s.sessionNumber >= fromSessionNumber && s.status !== 'COMPLETED')
    .sort((a, b) => a.sessionNumber - b.sessionNumber);

  if (targetSessions.length === 0) {
    return { updatedSessions: [], skippedDates: [] };
  }

  // استبعاد الجلسات المراد ترحيلها من مصفوفة الجلسات الحالية لتفريغ خاناتها السابقة
  const targetIds = new Set(targetSessions.map((s) => s.id));
  const otherCenterSessions = allCenterSessions.filter((s) => !targetIds.has(s.id));

  // توزيع التواريخ الجديدة للجلسات المتبقية
  const result = autoDistributeSessions(
    newStartDateStr,
    targetSessions.length,
    preferredDays,
    otherCenterSessions,
    capacitySettings,
    allowOverride,
    defaultCapacity
  );

  const updatedSessions = targetSessions.map((session, index) => {
    return {
      sessionId: session.id,
      oldDate: session.sessionDate,
      newDate: result.scheduledSessions[index].date
    };
  });

  return {
    updatedSessions,
    skippedDates: result.skippedDates
  };
}

File: medical_rehab_app/lib/core/scheduling_engine.dart
import '../models/session.dart';
import '../models/treatment_plan.dart';
import '../models/capacity_setting.dart';
import '../models/schedule_models.dart';

const int defaultMaxCapacity = 10;

const List<String> arabicDays = [
  'الأحد',
  'الإثنين',
  'الثلاثاء',
  'الأربعاء',
  'الخميس',
  'الجمعة',
  'السبت'
];

final Map<SchedulePattern, List<int>> patternDaysMap = {
  SchedulePattern.satMonWed: [6, 1, 3], // السبت، الإثنين، الأربعاء
  SchedulePattern.sunTueThu: [0, 2, 4], // الأحد، الثلاثاء، الخميس
  SchedulePattern.daily: [0, 1, 2, 3, 4, 6], // يومياً عدا الجمعة
  SchedulePattern.custom: [],
};

String formatDate(DateTime d) {
  final year = d.year.toString();
  final month = d.month.toString().padLeft(2, '0');
  final day = d.day.toString().padLeft(2, '0');
  return '$year-$month-$day';
}

DateTime parseDate(String dateStr) {
  final parts = dateStr.split('-');
  return DateTime(
    int.parse(parts[0]),
    int.parse(parts[1]),
    int.parse(parts[2]),
  );
}

String getArabicDayName(String dateStr) {
  final d = parseDate(dateStr);
  return arabicDays[d.weekday % 7]; // Dart DateTime.weekday: 1=Mon .. 7=Sun. d.weekday % 7: 7%7=0=Sun, 1%7=1=Mon
}

int getDayOfWeek(DateTime d) {
  return d.weekday % 7; // 0=Sunday, 1=Monday ... 6=Saturday
}

DayCapacitySummary calculateDayCapacity(
  String dateStr,
  List<Session> sessions,
  Map<String, DailyCapacitySetting> capacitySettings, [
  int defaultCapacity = defaultMaxCapacity,
]) {
  final d = parseDate(dateStr);
  final dayOfWeek = getDayOfWeek(d);
  final setting = capacitySettings[dateStr];

  // Friday is default holiday unless overridden
  final bool isDefaultWorkingDay = dayOfWeek != 5; // 5 = الجمعة
  final bool isWorkingDay = setting != null ? setting.isWorkingDay : isDefaultWorkingDay;
  final int maxCapacity = setting != null ? setting.maxCapacity : defaultCapacity;

  // Count active sessions (excluding CANCELLED)
  final bookedCount = sessions.where(
    (s) => s.sessionDate == dateStr && s.status != SessionStatus.cancelled,
  ).length;

  final percentage = maxCapacity > 0 ? ((bookedCount / maxCapacity) * 100).round() : 100;

  DayCapacityState status = DayCapacityState.available;
  if (!isWorkingDay) {
    status = DayCapacityState.holiday;
  } else if (bookedCount >= maxCapacity) {
    status = DayCapacityState.full;
  } else if (bookedCount >= maxCapacity * 0.8) {
    status = DayCapacityState.nearFull;
  }

  return DayCapacitySummary(
    date: dateStr,
    dayNameAr: arabicDays[dayOfWeek],
    dayOfWeek: dayOfWeek,
    bookedCount: bookedCount,
    maxCapacity: maxCapacity,
    isWorkingDay: isWorkingDay,
    percentage: percentage,
    status: status,
    notes: setting?.notes,
  );
}

AutoScheduleResult autoDistributeSessions({
  required String startDateStr,
  required int totalSessions,
  required List<int> preferredDays,
  required List<Session> existingSessions,
  required Map<String, DailyCapacitySetting> capacitySettings,
  bool allowOverride = false,
  int defaultCapacity = defaultMaxCapacity,
}) {
  if (totalSessions <= 0) {
    throw ArgumentError('عدد الجلسات يجب أن يكون أكبر من الصفر');
  }

  if (preferredDays.isEmpty) {
    throw ArgumentError('يجب تحديد يوم واحد على الأقل في نمط الجدولة');
  }

  final List<AutoSchedulePreviewItem> scheduledSessions = [];
  final List<SkippedDateItem> skippedDates = [];

  // Local simulated copy of sessions
  final simulatedSessions = List<Session>.from(existingSessions);

  DateTime currentDate = parseDate(startDateStr);
  const int maxDaysLookahead = 180;
  int daysScanned = 0;

  while (scheduledSessions.length < totalSessions && daysScanned < maxDaysLookahead) {
    final dayOfWeek = getDayOfWeek(currentDate);
    final dateStr = formatDate(currentDate);
    final dayNameAr = arabicDays[dayOfWeek];

    if (preferredDays.contains(dayOfWeek)) {
      final cap = calculateDayCapacity(dateStr, simulatedSessions, capacitySettings, defaultCapacity);

      if (!cap.isWorkingDay) {
        skippedDates.add(SkippedDateItem(
          date: dateStr,
          dayNameAr: dayNameAr,
          reason: cap.notes != null ? 'عطلة: ${cap.notes}' : 'عطلة أسبوعية أو رسمية',
        ));
      } else if (cap.bookedCount >= cap.maxCapacity && !allowOverride) {
        skippedDates.add(SkippedDateItem(
          date: dateStr,
          dayNameAr: dayNameAr,
          reason: 'اليوم ممتلئ بالكامل (${cap.bookedCount}/${cap.maxCapacity})',
        ));
      } else {
        final isOverrideNeeded = cap.bookedCount >= cap.maxCapacity;
        final sessionNum = scheduledSessions.length + 1;

        scheduledSessions.add(AutoSchedulePreviewItem(
          sessionNumber: sessionNum,
          date: dateStr,
          dayNameAr: dayNameAr,
          dayCapacity: cap.maxCapacity,
          dayBookedCount: cap.bookedCount + 1,
          isOverrideNeeded: isOverrideNeeded,
        ));

        simulatedSessions.add(Session(
          id: 'sim-$sessionNum',
          planId: 'sim-plan',
          patientId: 'sim-patient',
          sessionNumber: sessionNum,
          sessionDate: dateStr,
          status: SessionStatus.scheduled,
          createdAt: DateTime.now(),
        ));
      }
    }

    currentDate = currentDate.add(const Duration(days: 1));
    daysScanned++;
  }

  if (scheduledSessions.length < totalSessions) {
    throw StateError('لم نتمكن من جدولة جميع الجلسات. تم جدولة ${scheduledSessions.length} من أصل $totalSessions جلسة.');
  }

  final endDate = scheduledSessions.last.date;

  return AutoScheduleResult(
    scheduledSessions: scheduledSessions,
    skippedDates: skippedDates,
    endDate: endDate,
  );
}

ShiftResult shiftSubsequentSessions({
  required int fromSessionNumber,
  required String newStartDateStr,
  required List<int> preferredDays,
  required List<Session> allPlanSessions,
  required List<Session> allCenterSessions,
  required Map<String, DailyCapacitySetting> capacitySettings,
  bool allowOverride = false,
  int defaultCapacity = defaultMaxCapacity,
}) {
  final targetSessions = allPlanSessions
      .where((s) => s.sessionNumber >= fromSessionNumber && s.status != SessionStatus.completed)
      .toList()
    ..sort((a, b) => a.sessionNumber.compareTo(b.sessionNumber));

  if (targetSessions.isEmpty) {
    return ShiftResult(updatedSessions: [], skippedDates: []);
  }

  final targetIds = targetSessions.map((s) => s.id).toSet();
  final otherCenterSessions = allCenterSessions.where((s) => !targetIds.contains(s.id)).toList();

  final result = autoDistributeSessions(
    startDateStr: newStartDateStr,
    totalSessions: targetSessions.length,
    preferredDays: preferredDays,
    existingSessions: otherCenterSessions,
    capacitySettings: capacitySettings,
    allowOverride: allowOverride,
    defaultCapacity: defaultCapacity,
  );

  final updated = <ShiftedSessionItem>[];
  for (int i = 0; i < targetSessions.length; i++) {
    updated.add(ShiftedSessionItem(
      sessionId: targetSessions[i].id,
      oldDate: targetSessions[i].sessionDate,
      newDate: result.scheduledSessions[i].date,
    ));
  }

  return ShiftResult(
    updatedSessions: updated,
    skippedDates: result.skippedDates,
  );
}
